kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -678,61 +682,79 @@ impl Socket for StreamSocket {
+        if let Some(iface) = iface_to_poll {
+            iface.poll();
         }
 
-        // FIXME: Here we have only set the value of the option, without actually
-        // making any real modifications.
-        match_sock_option_ref!(option, {
-            tcp_no_delay: NoDelay => {
-                let no_delay = tcp_no_delay.get().unwrap();
-                options.tcp.set_no_delay(*no_delay);
-                state.set_raw_option(|raw_socket: &mut dyn RawTcpSetOption| raw_socket.set_nagle_enabled(!no_delay));
-            },
-            tcp_congestion: Congestion => {
-                let congestion = tcp_congestion.get().unwrap();
-                options.tcp.set_congestion(*congestion);
-            },
-            tcp_maxseg: MaxSegment => {
-                const MIN_MAXSEG: u32 = 536;
-                const MAX_MAXSEG: u32 = 65535;
-
-                let maxseg = tcp_maxseg.get().unwrap();
-                if *maxseg < MIN_MAXSEG || *maxseg > MAX_MAXSEG {
-                    return_errno_with_message!(Errno::EINVAL, "the maximum segment size is out of bounds");
-                }
-                options.tcp.set_maxseg(*maxseg);
-            },
-            tcp_window_clamp: WindowClamp => {
-                let window_clamp = tcp_window_clamp.get().unwrap();
-                let half_recv_buf = options.socket.recv_buf() / 2;
-                if *window_clamp <= half_recv_buf {
-                    options.tcp.set_window_clamp(half_recv_buf);
-                } else {
-                    options.tcp.set_window_clamp(*window_clamp);
-                }
-            },
-            _ => return_errno_with_message!(Errno::ENOPROTOOPT, "the socket option to be set is unknown")
-        });
-
         Ok(())
     }
 }
 
+fn do_tcp_setsockopt(
+    option: &dyn SocketOption,
+    options: &mut OptionSet,
+    state: &mut State,
+) -> Result<NeedIfacePoll> {
+    match_sock_option_ref!(option, {
+        tcp_no_delay: NoDelay => {
+            let no_delay = tcp_no_delay.get().unwrap();
+            options.tcp.set_no_delay(*no_delay);
+            state.set_raw_option(|raw_socket: &mut dyn RawTcpSetOption| raw_socket.set_nagle_enabled(!no_delay));
+        },
+        tcp_maxseg: MaxSegment => {
+            const MIN_MAXSEG: u32 = 536;
+            const MAX_MAXSEG: u32 = 65535;
+
+            let maxseg = tcp_maxseg.get().unwrap();
+            if *maxseg < MIN_MAXSEG || *maxseg > MAX_MAXSEG {
+                return_errno_with_message!(Errno::EINVAL, "the maximum segment size is out of bounds");
+            }
+            options.tcp.set_maxseg(*maxseg);
+        },
+        tcp_keep_idle: KeepIdle => {
+            const MIN_KEEP_IDLE: u32 = 1;
+            const MAX_KEEP_IDLE: u32 = 32767;
+
+            let keepidle = tcp_keep_idle.get().unwrap();
+            if *keepidle < MIN_KEEP_IDLE || *keepidle > MAX_KEEP_IDLE {
+                return_errno_with_message!(Errno::EINVAL, "the keep idle time is outside of bounds");
```

Comment:
```suggestion
                return_errno_with_message!(Errno::EINVAL, "the keep idle time is out of bounds");
```

Maybe we can unify the wording in line 722 and line 733.